### PR TITLE
Long-distance transmission

### PR DESCRIPTION
**Pull request by Hawk777**
These two commits between them should enable long-distance transmission of Digilines signals, including through mapblocks that are not anchored. See discussion in #35.

### CODE REVIEW

**Comment by raymoo** [comment]
what irc channel to I bother to get this reviewed

**Comment by MT-Modder** [comment]
@sofar !!!! lol

**Comment by sofar** [comment]
I don't see any existing review yet, @raymoo, can you at least review this as well?

It still has to be reviewed first. I'll look at that when I have time.

File: internal.lua [hunk 1]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
+end
 
-	local node = minetest.get_node(pos)
-	local spec = digiline:getspec(node)
-	if not spec then return end
+local function queue_empty(queue)
+	return queue.nextRead == queue.nextWrite
+end
 
+local function queue_push(queue, object)
```
Comment (raymoo) [inline internal.lua]: I would prefer the name `queue_enqueue` since "push" commonly refers to a stack operation.

File: internal.lua [hunk 2]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
+	local nextWrite = queue.nextWrite
+	queue[nextWrite] = object
+	queue.nextWrite = nextWrite + 1
+end
 
-	-- Effector actions --> Receive
-	if spec.effector then
-		spec.effector.action(pos, node, channel, msg)
-	end
+local function queue_pop(queue)
```
Comment (raymoo) [inline internal.lua]: I would prefer the name `queue_dequeue` since "pop" commonly refers to a stack operation.

File: internal.lua [hunk 3]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
+				spec.effector.action(curPos, node, channel, msg)
+			end
+
+			-- Cable actions --> Transmit
+			if spec.wire then
+				local rules = digiline:importrules(spec.wire.rules, node)
+				for _, rule in ipairs(rules) do
+					local nextPos = digiline:addPosRule(curPos, rule)
+					if digiline:rules_link(curPos, nextPos) then
+						local checkedID = tostring(nextPos.x) .. "_" .. tostring(nextPos.y) .. "_" .. tostring(nextPos.z)
```
Comment (raymoo) [inline internal.lua]: I know this is the original method, but why not hash_node_position? Likely faster than constructing a string with `..`s.

File: internal.lua [hunk 4]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
-	-- Cable actions --> Transmit
-	if spec.wire then
-		local rules = digiline:importrules(spec.wire.rules, node)
-		for _,rule in ipairs(rules) do
-			if digiline:rules_link(pos, digiline:addPosRule(pos, rule)) then
-				digiline:transmit(digiline:addPosRule(pos, rule), channel, msg, checked)
+function digiline:transmit(pos, channel, msg, checked)
+	digiline:vm_begin()
+	local queue = queue_new()
+	queue_push(queue, pos)
```
Comment (raymoo) [inline internal.lua]: see function comment

File: internal.lua [hunk 5]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
+			-- Cable actions --> Transmit
+			if spec.wire then
+				local rules = digiline:importrules(spec.wire.rules, node)
+				for _, rule in ipairs(rules) do
+					local nextPos = digiline:addPosRule(curPos, rule)
+					if digiline:rules_link(curPos, nextPos) then
+						local checkedID = tostring(nextPos.x) .. "_" .. tostring(nextPos.y) .. "_" .. tostring(nextPos.z)
+						if not checked[checkedID] then
+							checked[checkedID] = true
+							queue_push(queue, nextPos)
```
Comment (raymoo) [inline internal.lua]: see function comment

File: internal.lua [hunk 6]
```diff
@@ -63,28 +63,57 @@ function digiline:rules_link_anydir(output, input)
...
-		local rules = digiline:importrules(spec.wire.rules, node)
-		for _,rule in ipairs(rules) do
-			if digiline:rules_link(pos, digiline:addPosRule(pos, rule)) then
-				digiline:transmit(digiline:addPosRule(pos, rule), channel, msg, checked)
+function digiline:transmit(pos, channel, msg, checked)
+	digiline:vm_begin()
+	local queue = queue_new()
+	queue_push(queue, pos)
+	while not queue_empty(queue) do
+		local curPos = queue_pop(queue)
```
Comment (raymoo) [inline internal.lua]: see function comment

**Comment by Hawk777** [comment]
Reworked to fix the above comments.

File: util.lua [hunk 7]
```diff
@@ -65,3 +65,93 @@ function digiline:tablecopy(table) -- deep table copy
...
+--
+-- Outside a bulk-VoxelManipulator operation, if the mapblock is not loaded, it
+-- is pulled into the server’s main map data cache and then accessed from
+-- there.
+--
+-- Inside a bulk-VoxelManipulator operation, the operation’s VM cache is used.
+function digiline:get_node_force(pos)
+	if vm_cache then
+		return vm_get_node(pos)
+	else
```
Comment (DS-Minetest) [inline util.lua]: Because of the return this could be an `end`. Then the `end` down there wouldn't be needed and the code would be smaller.

**Comment by sofar** [comment]
If hawk can fix up that last comment then I'll merge this.

**Comment by Hawk777** [comment]
Done.

**Comment by sofar** [comment]
Merged.

**Pull request merged**